### PR TITLE
Adding equivalence and subset checks, including logical operations on DFAs (NOT, AND, OR)

### PR DESCRIPTION
**Pull request by ameesh-shah**
Credit to @mvcisback

### CODE REVIEW

File: dfa/utils.py [hunk 1]
```diff
@@ -38,3 +38,17 @@ def child_map(word_path):
     paths_ = tree.iddfs(max_depth=max_length, randomize=randomize)
 
     return (word for word, path in paths_ if end is None or path[-1] == end)
+
+def check_equivalence(dfa_a_, dfa_b_):
+    return check_is_subset(dfa_a_, dfa_b_) is None and check_is_subset(dfa_b_, dfa_a_) is None
```
Comment (mvcisback) [inline dfa/utils.py:43]: Could you make this function have output `None` is they are equiv and the counterexample otherwise?

File: dfa/utils.py [hunk 2]
```diff
@@ -38,3 +38,17 @@ def child_map(word_path):
     paths_ = tree.iddfs(max_depth=max_length, randomize=randomize)
 
     return (word for word, path in paths_ if end is None or path[-1] == end)
+
+def check_equivalence(dfa_a_, dfa_b_):
```
Comment (mvcisback) [inline dfa/utils.py:42]: Maybe rename to `find_equiv_counterexample`

File: dfa/utils.py [hunk 3]
```diff
@@ -38,3 +38,17 @@ def child_map(word_path):
...
+    return check_is_subset(dfa_a_, dfa_b_) is None and check_is_subset(dfa_b_, dfa_a_) is None
+
+def check_is_subset(dfa_, dfa_subset_candidate_):
```
Comment (mvcisback) [inline dfa/utils.py:45]: `find_subset_counterexample`

File: dfa/utils.py (hunk 2 above)
Comment (mvcisback) [reply dfa/utils.py:42]: also, are the trailing underscores intentional? Would prefer `dfa1` or `dfa_a` or even `left` and `right`

File: dfa/utils.py [hunk 4]
```diff
@@ -38,3 +38,17 @@ def child_map(word_path):
...
+    # first, get the complement of the non-candidate DFA
+    dfa12 = (~dfa_) & dfa_subset_candidate_
```
Comment (mvcisback) [inline dfa/utils.py:47]: Proposal: `dfa12` -> `should_be_empty`

File: dfa/utils.py (hunk 3 above)
Comment (mvcisback) [reply dfa/utils.py:45]: Proposal: `dfa_` -> `bigger` and `dfa_subset_candidata_` -> `smaller`, and switch order in arguments?

**Review by mvcisback — CHANGES_REQUESTED** [review]
Looks really good! Some small style tweaks and I'll happily merge.

I'd also really appreciate if you could add an example to the README.

**Comment by ameesh-shah** [comment]
addressing feedback in most recent commit

**Pull request merged**